src/query/functions/utils/heap.go

```diff
@@ -36,14 +38,29 @@ func maxHeapLess(i, j ValueIndexPair) bool {
 	if i.Val == j.Val {
 		return i.Index > j.Index
 	}
-	return i.Val < j.Val
+	return math.IsNaN(i.Val) && !math.IsNaN(j.Val) || i.Val < j.Val
 }
 
 func minHeapLess(i, j ValueIndexPair) bool {
 	if i.Val == j.Val {
 		return i.Index > j.Index
 	}
-	return i.Val > j.Val
+	return i.Val > j.Val || math.IsNaN(i.Val) && !math.IsNaN(j.Val)
+}
+
+func AscFloat64(i, j float64) bool {
+	return i < j || math.IsNaN(j) && !math.IsNaN(i)
+}
+
+func DescFloat64(i, j float64) bool {
+	return i > j || math.IsNaN(j) && !math.IsNaN(i)
+}
```

Comment:
nit: these should have more descriptive names like `GreaterWithNaNs`, and will need comments